Add PrometheusAgent to Getting Started docs (#5492)
prometheus-operator/prometheus-operator: Documentation/user-guides/getting-started.md

Start with the ServiceAccount, ClusterRole and ClusterRoleBinding:

```yaml mdox-exec="cat example/rbac/prometheus-agent/prometheus-service-account.yaml"
apiVersion: v1
kind: ServiceAccount
metadata:
  name: prometheus-agent
```

```yaml mdox-exec="cat example/rbac/prometheus-agent/prometheus-cluster-role.yaml"
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: prometheus-agent
rules:
- apiGroups: [""]
  resources:
  - services
  - endpoints
  - pods
  verbs: ["get", "list", "watch"]
- apiGroups: [""]
  resources:
  - configmaps
  verbs: ["get"]
- apiGroups:
  - networking.k8s.io
  resources:
  - ingresses
  verbs: ["get", "list", "watch"]
- nonResourceURLs: ["/metrics"]
  verbs: ["get"]
```

```yaml mdox-exec="cat example/rbac/prometheus-agent/prometheus-cluster-role-binding.yaml"
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: prometheus-agent
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: prometheus-agent
subjects:
- kind: ServiceAccount
  name: prometheus-agent
  namespace: default
```

> Lastly, we just need to deploy the Agent. Its spec is very similar to the Prometheus CR, with the difference the Agent CR has some functionalities disabled (rule evaluation for instance) to become more performant on the remote-write use-case.

Reviewer: ```suggestion
Lastly, we can deploy the Agent. The `spec` field is very similar to the Prometheus CRD but the features that aren't applicable to the agent mode (like alerting, retention, Thanos, ...) are not available.
```

Author: Somehow I missed this one all this time 😅. Should be good now

namespace. To select ServiceMonitors from other namespaces, you can update the
`spec.serviceMonitorNamespaceSelector` field of the Prometheus resource.

## Deploying Prometheus in Agent mode

Since Prometheus-Operator `v0.64.0` it is also possible to run Prometheus in Agent mode.

> As mentioned in [Prometheus's blog](https://prometheus.io/blog/2021/11/16/agent/), Prometheus Agent
> is a deployment model optimized for environments that expect that all collected data is being
> forwarded to a Long-term storage and local Prometheus don't not need storage at all.

Reviewer 2: ```suggestion
As mentioned in [Prometheus's blog](https://prometheus.io/blog/2021/11/16/agent/), Prometheus Agent
is a deployment model optimized for environments where all collected data is forwarded to 
a long-term storage solution (Cortex/Thanos/Mimir) and local Prometheus do not need storage or rule evaluation.
```

> To make it easier to upgrade from `v0.63.0` to `v0.64.0`, the operator is capable of running without
> permissions to the PrometheusAgent CR, but it won't reconcile PrometheusAgent resources if that's the case. So the first
> thing we need to do is make sure Prometheus-Operator has all the permissions necessary.

Reviewer 2: ```suggestion
To make it easier to upgrade from `v0.63.0` to `v0.64.0`, the operator is capable of running without
the permissions to reconcile the PrometheusAgent CR. So to use the new CR, make sure to update the operator permissions to allow reconciliations of the new CR.
```

    app.kubernetes.io/name: prometheus-operator
    app.kubernetes.io/version: 0.64.0
  name: prometheus-operator
rules:
- apiGroups:
  - monitoring.coreos.com
  resources:
  - alertmanagers
  - alertmanagers/finalizers
  - alertmanagers/status
  - alertmanagerconfigs
  - prometheuses
  - prometheuses/finalizers
  - prometheuses/status
  - prometheusagents
  - prometheusagents/finalizers
  - prometheusagents/status
  - thanosrulers
  - thanosrulers/finalizers
  - servicemonitors
  - podmonitors
  - probes
  - prometheusrules
  verbs:
  - '*'
- apiGroups:
  - apps
  resources:
  - statefulsets
  verbs:
  - '*'
- apiGroups:
  - ""
  resources:
  - configmaps
  - secrets
  verbs:
  - '*'
- apiGroups:
  - ""
  resources:
  - pods
  verbs:
  - list
  - delete
- apiGroups:
  - ""
  resources:
  - services
  - services/finalizers
  - endpoints
  verbs:
  - get
  - create
  - update
  - delete
- apiGroups:
  - ""
  resources:
  - nodes
  verbs:
  - list
  - watch
- apiGroups:
  - ""
  resources:
  - namespaces
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - networking.k8s.io
  resources:
  - ingresses
  verbs:
  - get
  - list
  - watch
```

> Now that we're past that, we just need to deploy the agent similarly to how we deployed Prometheus Server.

Reviewer 2: ```suggestion
The operator assigns different Kubernetes service accounts to Prometheus and Prometheus Agent. Because of this, it's also necessary to create a new service account for the Agent with the necessary permissions to scrape targets.
```

Reviewer 2: Yes, when we create the Statefulset
Server: https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/prometheus/server/statefulset.go#L501
Agent: https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/prometheus/agent/statefulset.go#L400

Reviewer 2: Sorry, I got confused with `Service` my bad :face_palm: here is an updated version
```
Similarly to Prometheus, Prometheus Agent will require permission to scrape targets. Because of this, we will create a new service account for the Agent with the necessary permissions to scrape targets.
```

Start with the ServiceAccount, ClusterRole and ClusterRoleBinding:

```yaml mdox-exec="cat example/rbac/prometheus-agent/prometheus-service-account.yaml"
apiVersion: v1
kind: ServiceAccount
metadata:
  name: prometheus-agent
```

```yaml mdox-exec="cat example/rbac/prometheus-agent/prometheus-cluster-role.yaml"
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: prometheus-agent
rules:
- apiGroups: [""]
  resources:
  - services
  - endpoints
  - pods
  verbs: ["get", "list", "watch"]
- apiGroups: [""]
  resources:
  - configmaps
  verbs: ["get"]
- apiGroups:
  - networking.k8s.io
  resources:
  - ingresses
  verbs: ["get", "list", "watch"]
- nonResourceURLs: ["/metrics"]
  verbs: ["get"]
```

```yaml mdox-exec="cat example/rbac/prometheus-agent/prometheus-cluster-role-binding.yaml"
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: prometheus-agent
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: prometheus-agent
subjects:
- kind: ServiceAccount
  name: prometheus-agent
  namespace: default
```

> Lastly, we just need to deploy the Agent. Its spec is very similar to the Prometheus CR, with the difference
> is that Agent mode has a lot of functionalities disabled to become more performant on the remote-write use-case.

Reviewer 2: ```suggestion
Lastly, we just need to deploy the Agent. Its spec is very similar to the Prometheus CR, with the difference the Agent CR has some functionalities disabled (rule evaluation for instance) to become more performant on the remote-write use-case.
```
